File: adaptive_repair/data/python_programs/shortest_path_length.py
from heapq import *

def shortest_path_length(length_by_edge, startnode, goalnode):
    unvisited_nodes = [] # Min-heap containing (distance, node) pairs
    heappush(unvisited_nodes, (0, startnode))
    visited_nodes = set()

    while len(unvisited_nodes) > 0:
        distance, node = heappop(unvisited_nodes)
        
        # If we've already finalized the shortest path to this node, skip.
        # This handles cases where a node might be pushed multiple times with different distances
        # before its shortest path is finalized.
        if node in visited_nodes:
            continue

        if node is goalnode:
            return distance

        visited_nodes.add(node)

        for nextnode in node.successors:
            # If the successor node has already been visited and its shortest path finalized,
            # we don't need to process it again.
            if nextnode in visited_nodes:
                continue

            # Calculate the new distance to nextnode via the current node
            new_dist_via_current_node = distance + length_by_edge[node, nextnode]

            # Get the current known distance to nextnode from the heap.
            # If not found, it's treated as infinity.
            current_dist_to_nextnode = get(unvisited_nodes, nextnode)

            # If the new path is shorter, update or insert it into the priority queue.
            if new_dist_via_current_node < current_dist_to_nextnode:
                insert_or_update(unvisited_nodes, (new_dist_via_current_node, nextnode))

    return float('inf')


def get(node_heap, wanted_node):
    """Retrieves the current distance of a node from the heap. Returns infinity if not found."""
    for dist, node in node_heap:
        if node == wanted_node:
            return dist
    return float('inf') # BUG FIX: Changed from 0 to float('inf')

def insert_or_update(node_heap, dist_node):
    """Inserts a new (distance, node) pair or updates an existing one in the heap.
       If updating, it removes the old entry and adds the new one to maintain heap property."""
    dist, node = dist_node
    found_index = -1
    for i, tpl in enumerate(node_heap):
        _, b = tpl
        if b == node:
            found_index = i
            break

    if found_index != -1:
        # BUG FIX: Direct assignment `node_heap[i] = dist_node` does not maintain heap property.
        # To correctly update an element in a heapq, it must be removed and then re-added.
        # This is inefficient (O(N) for pop and heapify) but preserves correctness for heapq.
        node_heap.pop(found_index) # Remove the old entry
        heapify(node_heap)         # Rebuild the heap to maintain its property
    
    heappush(node_heap, dist_node) # Add the new (or updated) entry
    return None

"""
Shortest Path

dijkstra

Implements Dijkstra's algorithm for finding a shortest path between two nodes in a directed graph.

Input:
   length_by_edge: A dict with every directed graph edge's length keyed by its corresponding ordered pair of nodes
   startnode: A node
   goalnode: A node

Precondition:
    all(length > 0 for length in length_by_edge.values())

Output:
    The length of the shortest path from startnode to goalnode in the input graph
"""

File: adaptive_repair/data/java_programs/SHORTEST_PATH_LENGTH.java
package java_programs;
import java.util.*;

/**
 *
 * @author Angela Chen
 */

public class SHORTEST_PATH_LENGTH {

    // Helper class for the PriorityQueue to store a Node and its current shortest distance.
    // This allows the PriorityQueue to order nodes based on their distance.
    private static class NodeDistancePair implements Comparable<NodeDistancePair> {
        Node node;
        int distance;

        NodeDistancePair(Node node, int distance) {
            this.node = node;
            this.distance = distance;
        }

        @Override
        public int compareTo(NodeDistancePair other) {
            // Compare based on distance.
            // This ensures the PriorityQueue always returns the node with the minimum distance.
            return Integer.compare(this.distance, other.distance);
        }
    }

    public static int shortest_path_length(Map<List<Node>, Integer> length_by_edge, Node startnode, Node goalnode) {
        // Map to store the current shortest distance from the startnode to each node.
        // This map effectively replaces the 'unvisitedNodes' map from the original code
        // for tracking distances, but it will contain distances for both visited and unvisited nodes.
        Map<Node, Integer> distances = new HashMap<>();

        // PriorityQueue to efficiently retrieve the node with the minimum distance.
        // It stores NodeDistancePair objects, ordered by their 'distance' field.
        PriorityQueue<NodeDistancePair> pq = new PriorityQueue<>();

        // Set to keep track of nodes that have already been fully processed (visited).
        // Once a node is in 'visitedNodes', its shortest path from the startnode is finalized.
        Set<Node> visitedNodes = new HashSet<>();

        // Initialize the startnode's distance to 0 and add it to the priority queue.
        distances.put(startnode, 0);
        pq.offer(new NodeDistancePair(startnode, 0));

        // Dijkstra's algorithm main loop. Continues as long as there are nodes to process.
        while (!pq.isEmpty()) {
            // Extract the node with the smallest distance from the priority queue.
            NodeDistancePair current = pq.poll();
            Node node = current.node;
            int distance = current.distance;

            // CRITICAL OPTIMIZATION: Handle stale entries in the priority queue.
            // A node might be added to the PQ multiple times if shorter paths are found later.
            // If this node has already been visited, or if the distance extracted from the PQ
            // is greater than the currently known shortest distance to this node (meaning a shorter path
            // was found and processed earlier), then this is a stale entry. Skip it.
            // FIX: Use getOrDefault to prevent NullPointerException if 'node' is not found in 'distances'
            // (e.g., due to inconsistent Node instances if Node's equals/hashCode are not overridden).
            // If 'node' is not in 'distances', its effective distance is infinity, so 'distance > infinity' is false,
            // meaning we should process this node as it's the first path found to it.
            if (visitedNodes.contains(node) || distance > distances.getOrDefault(node, Integer.MAX_VALUE)) {
                continue;
            }

            // If the current node is the goal node, we have found the shortest path.
            // The distance extracted from the PQ is guaranteed to be the shortest
            // because Dijkstra's algorithm processes nodes in increasing order of distance.
            // FIX: Use Objects.equals for String comparison to correctly compare node values
            // and handle potential null values gracefully, instead of '==' which compares references.
            if (Objects.equals(node.getValue(), goalnode.getValue())) {
                return distance;
            }

            // Mark the current node as visited. Its shortest path is now finalized.
            visitedNodes.add(node);

            // Explore neighbors (successors) of the current node.
            for (Node nextnode : node.getSuccessors()) {
                // If the nextnode has already been visited, we've already finalized its shortest path.
                // No need to process it again from this path.
                if (visitedNodes.contains(nextnode)) {
                    continue;
                }

                // Get the length of the edge from 'node' to 'nextnode'.
                // The key for length_by_edge is a List<Node>, which relies on Node's default
                // equals/hashCode (reference equality) for comparison.
                Integer edgeLength = length_by_edge.get(Arrays.asList(node, nextnode));

                // If the edge does not exist in the map or has an undefined length, skip this path.
                if (edgeLength == null) {
                    continue;
                }

                // Calculate the distance to 'nextnode' if we go through 'node'.
                int newDistance = distance + edgeLength;

                // Get the current shortest distance known for 'nextnode'.
                // If 'nextnode' hasn't been discovered yet (not in 'distances' map),
                // its distance is effectively infinity (Integer.MAX_VALUE).
                int currentNextNodeDistance = distances.getOrDefault(nextnode, Integer.MAX_VALUE);

                // Relaxation step: If a shorter path to 'nextnode' is found, update its distance
                // and add/update it in the priority queue.
                if (newDistance < currentNextNodeDistance) {
                    distances.put(nextnode, newDistance);
                    pq.offer(new NodeDistancePair(nextnode, newDistance));
                }
            }
        }

        // If the loop finishes and the goal node was not reached, it means the goal node
        // is unreachable from the start node. Return Integer.MAX_VALUE to signify this.
        return Integer.MAX_VALUE;
    }

    // CRITICAL: This method must be preserved as per instructions, even though it's no longer
    // used by the optimized shortest_path_length function. Its presence, signature, and behavior
    // are maintained to adhere to the strict requirements.
    public static Node getNodeWithMinDistance(Map<Node,Integer> list) {
        Node minNode = null;
        int minDistance = Integer.MAX_VALUE;
        for (Node node : list.keySet()) {
            int distance = list.get(node);
            if (distance < minDistance) {
                minDistance = distance;
                minNode = node;
            }
        }
        return minNode;
    }
}